src/Servers/Kestrel/Transport.Libuv/test/ListenerPrimaryTests.cs

```diff
@@ -270,44 +237,73 @@ public async Task PipeConnectionsWithWrongMessageAreLoggedAndIgnored()
             Assert.Contains("Bad data", errorMessage.Exception.ToString());
         }
 
-        private static async Task AssertResponseEventually(
-            Uri address,
-            string expected,
-            string[] allowed = null,
-            int maxRetries = 100,
-            int retryDelay = 100)
+
+        private static async Task AssertRoundRobin(Uri address, ListenerPrimary listenerPrimary, ListenerSecondary listenerSecondary, ListenerContext currentListener, Task<LibuvConnection> expected = null, int connections = 4)
         {
-            for (var i = 0; i < maxRetries; i++)
+            for (int i = 0; i < connections; i++)
             {
-                var response = await HttpClientSlim.GetStringAsync(address);
-                if (response == expected)
+                if (currentListener == listenerPrimary)
                 {
-                    return;
+                    expected ??= listenerSecondary.AcceptAsync().AsTask();
+                    currentListener = listenerSecondary;
                 }
-
-                if (allowed != null)
+                else
                 {
-                    Assert.Contains(response, allowed);
+                    expected ??= listenerPrimary.AcceptAsync().AsTask();
+                    currentListener = listenerPrimary;
                 }
 
-                await Task.Delay(retryDelay);
-            }
+                using var socket = await HttpClientSlim.GetSocket(address);
+
+                await using var connection = await expected.DefaultTimeout();
```

Comment:
Yeah, you did take some artistic license on this test. It would be nice if there was a quicker way we could verify HttpClientSlim actually has the same socket listenerPrimary does. Before, the "Primary" vs. "Secondary" responses gave you a quick error instead of this test failing with a timeout.

Reply:
It can't though. We need to verify what's in the queues, there's no way to have different "applications" for different listeners, the fact that it worked that way before was a bit of a fluke because we had a different dispatcher per listener and we took advantage of that.

Reply:
You could write "Primary" or "Secondary" to the LibuvConnection after accepting it, no?

Reply:
Sure, I can do that but the real test is the fact that it actually is en-queued in the right queue, not the writing to the queue.